### PR TITLE
Refresh Cody Pro status on editor focus change

### PR DESCRIPTION
**Pull request by SuperAuguste**
Addresses #1992 for VSCode and enables other editors to address it.

## Test plan

Tested locally.

### CODE REVIEW

File: vscode/src/main.ts [hunk 1]
```diff
@@ -538,6 +538,19 @@ const register = async (
     authProvider.addChangeListener(() => updateAuthStatusBarIndicator())
     updateAuthStatusBarIndicator()
 
+    vscode.window.onDidChangeWindowState(async ws => {
```
Comment (kalanchan) [inline vscode/src/main.ts:541]: how often would this get fired? Is this similar to something like `onBlur` in the DOM?

File: agent/src/agent.ts [hunk 2]
```diff
@@ -386,6 +386,19 @@ export class Agent extends MessageHandler {
             throw id
         })
 
+        this.registerRequest('graphql/currentUserIsPro', async () => {
```
Comment (kalanchan) [inline agent/src/agent.ts:389]: do we need another graphql query? Unless i'm mistaken there should already be a graphql query to check account status, or maybe that was just on the VSC side

We were using this query on VSC
```{
  currentUser {
    codyProEnabled
  }
```

**Review by kalanchan — COMMENTED** [review]
I think the approach here is good, I just have a couple questions around when the check would get fired. 

We need be able to handle scenarios where
1. user starts in the web, upgrades, and goes back to editor
2. user starts in the editor, upgrades, and goes back to the editor

File: lib/shared/src/sourcegraph-api/graphql/queries.ts [hunk 3]
```diff
@@ -251,3 +251,13 @@ export const EVALUATE_FEATURE_FLAG_QUERY = `
         evaluateFeatureFlag(flagName: $flagName)
     }
 `
+
+export const CURRENT_USER_HAS_CODY_PRO = `
```
Comment (abeatrix) [inline lib/shared/src/sourcegraph-api/graphql/queries.ts:255]: This looks similar to the `CURRENT_USER_ID_AND_VERIFIED_EMAIL_AND_CODY_PRO_QUERY`query, we can just use that instead?

File: lib/shared/src/sourcegraph-api/graphql/client.ts [hunk 4]
```diff
@@ -332,6 +337,19 @@ export class SourcegraphGraphQLAPIClient {
...
 
+    public async getCurrentUserHasCodyPro(): Promise<
+        { id: string; codyProEnabled: boolean; codyProEnabledAt: Date } | Error
+    > {
+        return this.fetchSourcegraphAPI<APIResponse<CurrentUserHasCodyProResponse>>(CURRENT_USER_HAS_CODY_PRO, {}).then(
+            response =>
+                extractDataOrError(response, data =>
+                    data.currentUser
+                        ? { ...data.currentUser }
+                        : new Error('current user not found with verified email and cody pro')
```
Comment (abeatrix) [inline lib/shared/src/sourcegraph-api/graphql/client.ts:348]: The `CURRENT_USER_HAS_CODY_PRO` query doesn't tell us if the user has verified email or not, maybe we can just use the `getCurrentUserIdAndVerifiedEmailAndCodyPro` function below for the same purpose?

**Review by abeatrix — COMMENTED** [review]
left comments and questions in line. I don't think this would work if the users do the sign up process in another device, but i guess that's an edge case for another time? 😜

**Comment by kalanchan** [comment]
> left comments and questions in line. I don't think this would work if the users do the sign up process in another device, but i guess that's an edge case for another time? 😜

I think the idea is that it doesn't matter where the user initiates the upgrade/downgrade, whenever they go back to the editor and click around it should fire a check. Is that right @SuperAuguste?


but yes, we need to handle both scenarios: 

1. user starts in the web, upgrades/downgrades, and goes back to editor
2. user starts in the editor, upgrades/downgrades, and goes back to the editor

File: vscode/src/main.ts (hunk 1 above)
Comment (SuperAuguste) [reply vscode/src/main.ts:541]: Whenever a user comes back to a VSCode window after clicking off it. :)

File: lib/shared/src/sourcegraph-api/graphql/queries.ts (hunk 3 above)
Comment (SuperAuguste) [reply lib/shared/src/sourcegraph-api/graphql/queries.ts:255]: Sure!

**Review by abeatrix — APPROVED** [review]
yay

**Comment by toolmantim** [comment]
Nice one @SuperAuguste 🎉

**Review by vdavid — APPROVED** [review]
Nice, likey!

**Comment by SuperAuguste** [comment]
Glad this helps a little!

**Pull request merged**